x/bank/keeper/keeper.go

```diff
@@ -183,6 +183,14 @@ func (k BaseKeeper) DelegateCoins(ctx sdk.Context, delegatorAddr, moduleAccAddr
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amt.String())
 	}
```

Comment:
Asking to learn: Don't we need to add call to `BeforeSend` hook as well in `subUnlockedCoins`?

Reply:
If we call it in subUnlockedCoins as well, wouldn't it get called twice?

Reply:
It seems that we add the `BeforeSend` before every occurrence of `subUnlockedCoins` but `BurnCoins`:
https://github.com/osmosis-labs/cosmos-sdk/blob/8fe17b10bca5e8c1ad57e45ed6515daaeed57dbf/x/bank/keeper/keeper.go#L564

What do you think about introducing a `subUnlockedCoinsWithBeforeSendHook` that would simply apply the hook and then call `subUnlockedCoins`. We could use this function in all occurrences other than in `BurnCoins`

IMO that would make it easier to track where the hook is used

Reply:
But `BurnCoins` isn't really a Send though, right?

Reply:
Yes, we should not need `BeforeSend` in `BurnCoins`.

My point was that since the general trend is that we call `BeforeSend` before `subUnlockedCoins` in all cases but `BurnCoins`, it might make sense to group `BeforeSend` and `subUnlockedCoins` into a single function `subUnlockedCoinsWithBeforeSendHook`.

So `BurnCoins` would be kept as is (calling `subUnlockedCoins`) while all other logic would switch to using `subUnlockedCoinsWithBeforeSendHook`.

This is a minor suggestion to reduce code duplication and group related logic so I wouldn't spend too much time on it. The current design looks good so we can keep it if you don't see much value in the grouping I suggested.